test/lib/Transforms/TestListenerPasses.cpp

```diff
@@ -0,0 +1,94 @@
+//===- TestListenerPasses.cpp - Test passes with listeners ----------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "Transforms/Listener.h"
+#include "Transforms/ListenerCSE.h"
+#include "Transforms/ListenerGreedyPatternRewriteDriver.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+
+namespace {
+
+/// The test listener prints stuff to `stdout` so that it can be checked by lit
+/// tests.
+struct TestListener : public RewriteListener {
+  void notifyOperationReplaced(Operation *op, ValueRange newValues) override {
+    llvm::outs() << "REPLACED " << op->getName() << "\n";
+  }
+  void notifyOperationRemoved(Operation *op) override {
+    llvm::outs() << "REMOVED " << op->getName() << "\n";
+  }
+};
+
+struct TestCanonicalizePass : public PassWrapper<TestCanonicalizePass, Pass> {
```

Comment:
Nit: I wouldn't call it "test canonicalize" to (a) avoid confusion with the "main" canonicalize and (b) make it clear that it is testing the rewriter infra rather than canonicalization patterns.